_posts/2019-05-22-search-is-a-team-work.md

```diff
@@ -0,0 +1,132 @@
+---
+typora-copy-images-to: ..\img\posts\search-is-a-team-work
+typora-root-url: ..
+layout: post
+header_image: /img/posts/search-is-a-team-work/header.jpg
+title: "Search is a Team Work"
+author: Mauro Servienti
+synopsis: "Designing a search functionality for a distributed system is challenging. There is the coupling threat, on the one hand, and on the other we cannot to sacrifice efficiency. There are a lot of analogies with what happens during a race car pit stop."
+tags:
+- SOA
+- Search
+category: soa-search
+---
+
+In last installment, [The Quest for a better Search](<https://milestone.topics.it/soa-search/2019/05/15/the-quest-for-a-better-search.html>), [Tomek](https://twitter.com/Masternak) and myself literally scratched the surface by looking at the high level problem that searching in a distributed system brings to the table.
+
+Let's use the Hotel Reservations scenario as a way to identify some requirements:
+
+* Users need to be able to search hotels by location, .e.g. a city.
+* The search criteria can contain a check-in date, and/or a check-out date to filter out unavailable hotels.
+* Results can be further filtered by:
+  * amenities, such as breakfast included, or a spa. 
+  * stars
+  * guests reviews and ratings
+
+By looking at the above requirements and doing a very rough, and highly unsuggested, boundaries analysis, we could identify that:
+
+* Registry owns:
+  * the hotel location
+  * amenities
+  * stars
+* Reservations owns availability
+* Marketing owns guests reviews and ratings
+
+## A naïf design
+
+If we look at the above requirements, as they are listed, in theory we could design a distributed search process like the following (by applying some [ViewModel Composition techniques](https://milestone.topics.it/categories/view-model-composition)):
+
+* Registry will handle the first part of the query returning the list of hotel IDs matching the `location` criteria
+* Returned IDs are then handover to Reservations that filters out, based on selected dates, IDs matching hotels not available
+* Finally IDs are passed to Marketing, and again Registry, to execute a [Faceted search](https://en.wikipedia.org/wiki/Faceted_search) and build facets results.
+
+It works, still is very inefficient. It would be like if the engineers from the header image of this article were to move the race car to different locations to change tyres or to refuel it. Very inefficient.
```

Comment:
Should this paragraph state explicitly that too much data is being moved between storage and composition process or even between processes via remoting? - to drive the point home?

Reply:
Yes, I’ll reword it a bit.

Reply:
Reworded. I’ll go ahead and merge this. We can always improve later, and there is still an editorial review pending.